### ISSUE TITLE
Inverted link pattern does not satisfy accessibility contrast requirements against adjacent text

### ISSUE THREAD

**Issue by lyubomir-popov** [canonical-web-and-design/vanilla-framework#3350]
## Pattern to amend

[Inverted link](https://vanillaframework.io/docs/examples/patterns/links/links-inverted)

The inverted link pattern does not satisfy accessibility contrast requirements against adjacent text. ([Source](https://webaim.org/blog/wcag-2-0-and-link-colors/))
![image](https://user-images.githubusercontent.com/2741678/95579311-4b6a5d00-0a2d-11eb-9f76-afc0035090fd.png)

**Comment by lyubomir-popov** [comment]
Starting with the link colour on white (`#06c`) in order to satisfy both conditions:
- > 4.5 contrast to background
- > 3.0 contrast to surrounding text (#fff)

I've arrived at the following colour. It has a requirement though: the background should not be ligher than `#2d2d2d`, sowe'd have to reconsider use of this on e.g. `#333`

![image](https://user-images.githubusercontent.com/2741678/110443641-0c136b80-80b4-11eb-828f-a55f23e82b2f.png)

**Comment by lyubomir-popov** [comment]
- Agreed to add mandaory chevron to inverted link, keep it white, and mention in the documentation that it is meant to be used on takeovers and places where we can't set stringent contrast requirements on the background.
- Introduce the blue described above, create example, and mention in the documentation that it should not be placed on surfaces ligher than #2d2d2d, as that would make it impossible to simultaneously maintain 3.0 contrast to surrounding text and 4.5 contrast to background colour

**Comment by anthonydillon** [comment]
> Agreed to add mandaory chevron to inverted link, keep it white, and mention in the documentation that it is meant to be used on takeovers and places where we can't set stringent contrast requirements on the background.

We cannot make this mandatory but we could recommend it. Takeovers are the most bespoke parts of any website and so the background is likely to be any colour or gradient.

> Introduce the blue described above, create example, and mention in the documentation that it should not be placed on surfaces ligher than #2d2d2d, as that would make it impossible to simultaneously maintain 3.0 contrast to surrounding text and 4.5 contrast to background colour

So this would change the inverted pattern from white to blue?

**Comment by lyubomir-popov** [comment]
> We cannot make this mandatory but we could recommend it

I was thinking using the selector to add a pseudo element with a chevron, same as the external link. Would that not work?


> So this would change the inverted pattern from white to blue?

No, he inverted link stays white. We need to work out  the best way to do the on-dark blue link - maybe either a child of .is dark, or .is-on-dark directly on the link?

**Comment by anthonydillon** [comment]
> I was thinking using the selector to add a pseudo element with a chevron, same as the external link. Would that not work?

It would work but would be a breaking change. why does the inverted link need a suffix?

> No, he inverted link stays white. We need to work out the best way to do the on-dark blue link - maybe either a child of .is dark, or .is-on-dark directly on the link?

Are you trying to stop people using the blue link on dark backgrounds or provide the ability for it to just work?

**Comment by lyubomir-popov** [comment]
> It would work but would be a breaking change. why does the inverted link need a suffix?

To comply with AA accessibility requirements. Link colour needs to have 3:1 contrast with sourrounding text. 

More info: 

https://webaim.org/blog/wcag-2-0-and-link-colors/
https://www.w3.org/TR/2008/NOTE-WCAG20-TECHS-20081211/G183

**Comment by anthonydillon** [comment]
> Designator

I wonder if a chevron would be regarded as enough to be a designator?

> Blue

Ok thanks

**Comment by lyubomir-popov** [comment]
> I wonder if a chevron would be regarded as enough to be a designator?

It is a strong convention - many sites use it as an indication that clicking will take you somewhere, i.e. it is a link. E.g. apple and ios use it exclusively (albeit in conjunction with blue colour). 

I'm just trynig to find a solution to an accessibility failure, and I believe underlines were present at some point and someone requested hey be removed, so feels like it might be perceived as a regression.

**Comment by anthonydillon** [comment]
@lyubomir-popov @spencerbygraves My question is, have we considered using a underline on a inverted link instead of the chevron because the chevron could leads to issues such as double chrevons which would need manual updates to fix. The inverted link rarely used and so I think would be ok to adopt the underline when others don't.

**Comment by SirSamTumless** [comment]
Waiting for 3.0 release as it would be a breaking change, but needs a bit more investigation from @lyubomir-popov

**Comment by lyubomir-popov** [comment]
Talking with @spencerbygraves and @anthonydillon, we decided to postpone this to the 3.0 release as introducing a chevron would cause double chevrons in places that already manually append a single guillemet (french quote) at the end of links.

**Comment by bartaz** [comment]
- chevron added with CSS will wrap, we used to have it and we needed to remove
- investigate if we can make it not wrap
- or have a blue link (this has it's own limitations)

**Comment by bartaz** [comment]
It seems that chevron on end of the link will wrap in some cases if added in CSS.
If possible we should revisit how inverted links are styled.

**Comment by clagom** [comment]
@lyubomir-popov to create a mockup of the example with the chevron (the option that the WG chose as a solution).

**Comment by clagom** [comment]
WG: @lyubomir-popov to discuss this during Design Assembly to agree on a compromise

**Comment by marinacastejon** [comment]
Hi, 
We discussed about this in the accessibility guild. 
[This example was shown](https://charmed-osm.com/osm-hackfests) (see bottom of the page)
We agreed on this being not compliant and wanted to shared a good resource that summarises well the WCAG guideline: [Yale university](https://usability.yale.edu/web-accessibility/articles/links). Non-colour indicators include underlining and also adding an icon, so in principle the chevron should be accessible. To be discussed what is the best compromise.

**Comment by clagom** [comment]
@marinacastejon @lyubomir-popov can you raise this issue during the next design assembly please?
CC @amylily1011 as part of the work you did

**Comment by lyubomir-popov** [comment]
Added to the agenda https://docs.google.com/document/d/160W5VxwUPszuNzV1pzi5f9XO6Yl-RVbcZG3faCq4hsI/edit#

**Comment by lyubomir-popov** [comment]
Discussed in today's design assembly:

![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/a2e9e291-7501-40d1-abda-65b54bc48766)

**Comment by lyubomir-popov** [comment]
Decision from design assembly:
- reintroduce underline on all links in Vanilla
- Accessible blue on dark grey, + underline:
- White underlined links on varying backgreounds (gradients, images in hero elements)
Next steps: apply the above rules to a few pages, reconvene to finalise decision

![image](https://user-images.githubusercontent.com/2741678/155127031-45fc26d7-4b5a-4a0e-b4b7-c16f7173dc96.png)

**Comment by lyubomir-popov** [comment]
Some thoughts after testing the above decision on ubuntu.com:

1) Not sure whether this is due to the typeface, but underlines on ubuntu (especially light, thin) are surprisingly thick and unbalanced compared to the font-weight:

![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/b4cacb17-549b-4489-b8d4-d9c81a4322d9)

2) ubuntu.com makes heavy use of links. So much so that sometimes over 50% of the copy is links. In those cases I think the readability of the text suffers because of the thick underlines:
![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/9da7bbc5-8320-4cd1-ad3a-9fa6b8cab9c1)

![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/e7156d81-8cbc-4cce-9430-f500fd8d74a9)

The same holds for headings, especially when multiple links break up the texture of the regular text:
![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/dfc5faaa-7165-45c2-8e00-6862fa429c8e)

Based on the above, my recommendation would be to limit the underline exclusively to where we need it for accessibility reasons. 
So instead of compromising the readability of all links across the board, we limit this compromise to usecases where the 3:1 contrast requirement against background and adjacent text forces us to use an underline. This woul be e.g. gradients, on top of images, or backgrounds lighter than #2d2d2d, which is the threshold beyond which a blue link colour cannot simultaneously pass 3:1 contrast against white adjacent text and dark background.

So:
By default, links (in the dark theme only) are white and underlined:
![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/a5044f7a-fb36-458a-bddb-17062edb2fa6)

On hover they turn blue (contrast 3:1 against adjacent white, precise hex #3097FF). This is as light as we can make it without breaching the contrast ratio.

visited links take the visited state from the white theme and nudge the colour until it satisfies contrast ratio requirements against white: #B973FF
![image.png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/ef726fce-ece0-4b0c-9404-afd5a2bdfd36)

whole page with t he problems above: 
![ubuntu.com_ (5).png](https://images.zenhubusercontent.com/59d2496bb0222d5de479ea17/c8d94568-b8bb-4b1a-8902-edd8b0fb3b92)

**Comment by clagom** [comment]
That underline is really something we need to fix in the new font @marcushas @waynecrosby, how come we didn't have this issue before?

**Comment by waynecrosby** [comment]
@clagom It is a common pattern for the underline to be slightly heavier, although the Thin is particularly heavy and more obvious. I think the amount of links makes it more obvious too. We've contacted Dalton Maag to get their input and will let you know their thoughts.

**Comment by lyubomir-popov** [comment]
Update:  this can be controlled via the `text-decoration-thickness` css property, although it would be good for the font metric to be updated (`text-decoration-thickness: from-font;` is the default it seems). Here is an updated example:
![ubuntu com_ (6)](https://user-images.githubusercontent.com/2741678/157216129-8968881e-2dff-44b8-b937-25cefa80f9a2.png)

**Comment by lyubomir-popov** [comment]
Update: there's also a property for underline offset, I've used it to distance the underline from the text, which improves legibility to a certain extent. 

Still, the excessive use of links is something to think about in a separate conversation.

As to the choice here, the options are presented in this zeplin project: https://app.zeplin.io/project/62273eecef3ac2b9258f9440
Please visit the link and comment on which approach you prefer - 
a) apply the underline across the board
b) limit the underline to the dark theme

**Comment by cristinadresch** [comment]
@pmahnke any thoughts on next steps for this, did you manage to show the demo link around?

**Comment by pmahnke** [comment]
I haven't really shown it around too much, just Thibaut (who said it was 'like the 90's web'). Some things I noticed.

![image](https://user-images.githubusercontent.com/441217/158801235-85053cd0-73a0-4bc3-a18b-3b1d2dbfb338.png)
I know it is a quick demo, but I wouldn't like the underline in some situations and I am not sure how that would mess things up.

![image](https://user-images.githubusercontent.com/441217/158801446-6048470c-45dd-4a22-a8db-29c7b550cfca.png)
This looks bad as well, 

![image](https://user-images.githubusercontent.com/441217/158801665-326364b9-7c5c-463b-852d-459dcba6f093.png)
If you remove `is-bordered` it looks better, but then using that pattern is a bit more dangerous.

![image](https://user-images.githubusercontent.com/441217/158802259-0885327b-6ea5-477c-b043-95e5517d0f15.png)
Headings don't look great either... and docs tocs...

Basically, I don't love them in lots of scenarios, so I don't think it really works.

**Comment by lyubomir-popov** [comment]
Thanks @pmahnke, I think that settles it right @clagom? We limit this to dark backgrounds, leaving 99% of pages unaffected.

**Comment by lyubomir-popov** [comment]
Sure, I'll make some mockups

**Comment by lyubomir-popov** [comment]
(Trying an email reply as github is having issues, hope this gets displayed
correctly).

@clagom The alternative we discussed in the WG UX/UI meetings was as
follows:
1) Limit underlines to the dark theme.  On dark backgrounds, apply both
colour and underline for maximum clarity:

![image](
https://user-images.githubusercontent.com/2741678/158828189-744003e8-89af-4cbb-8024-13c292861e11.png
)

2) In cases where the background colour is too light to ensure 4.5:1
contrast between link colour (the dark theme blue link colour to be
precise), we drop the blue and use the underline only:

![image](
https://user-images.githubusercontent.com/2741678/158828414-44eca3e4-9724-4ed9-b7c9-598fa4967bfe.png
)

Within the dark theme, in case of structures that are reasonably recognised
as interactive, e.g. menus, footers, etc, use non-underlined white text. To
avoid accessibility breaches, non-clickable text should not be mixed within
the menus.

![image](
https://user-images.githubusercontent.com/2741678/158829309-f6932974-f2c0-4126-b45c-96833a779215.png
)

If we need to mix clickable and non clickable text, we fall back to option
[2] above.

Updated zeplin link:
https://zpl.io/GEkPBpm

**Comment by lyubomir-popov** [comment]
@clagom The alternative we discussed in the WG UX/UI meetings was as follows:
1) Limit underlines to the dark theme.  On dark backgrounds, apply both colour and underline for maximum clarity:
 
![image](https://user-images.githubusercontent.com/2741678/158828189-744003e8-89af-4cbb-8024-13c292861e11.png)

2) In cases where the background colour is too light to ensure 4.5:1 contrast between link colour (the dark theme blue link colour to be precise), we drop the blue and use the underline only:

![image](https://user-images.githubusercontent.com/2741678/158828414-44eca3e4-9724-4ed9-b7c9-598fa4967bfe.png)

Within the dark theme, in case of structures that are reasonably recognised as interactive, e.g. menus, footers, etc, use non-underlined white text. To avoid accessibility breaches, non-clickable text should not be mixed within the menus. 

![image](https://user-images.githubusercontent.com/2741678/158829309-f6932974-f2c0-4126-b45c-96833a779215.png)

If we need to mix clickable and non clickable text, we fall back to option [2] above.

Updated zeplin link:
https://zpl.io/1ykAlAm

**Comment by clagom** [comment]
Thanks @lyubomir-popov. It seems hard to find a simpler solution, unfortunately, without compromising quality. I hope Vanilla can scale to be this modular and that hopefully the QA of our process is enough to catch inconsistencies.

**Comment by lyubomir-popov** [comment]
It's a complex set of usecases and accessibility requirements, which is why we originally wanted one treatment for both light and dark themes, in order to simplify the logic. With that option removed, it is indeed hard to simplify the rules around link styling any further. 

I see no scalability issues - these are just classes that get applied on request, nothing is automatic, but we do need to watch out for inconsistencies.

@clagom can we consider this ready for development and close this issue?

**Comment by clagom** [comment]
Yes that makes sense @lyubomir-popov, thanks for your work, helped a lot

**Comment by lyubomir-popov** [comment]
@anthonydillon the inverted link component will have 2 flavours:

#### When to use:
When links are surrounded by regular, non-clickable text. Depending on the background, there are 2 choises:
- underlined blue link - the default, which is intended for links on backgrounds #2d2d2d or lighter
- underlined white link - on backgrounds lighter thant #2d2d2d, where it is impossible to obtain a blue colour that simultaneously has 3:1 contrast with surrounding white text, and at the same time, 4.5:1 contrast with the background as per WCAG rule [1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) and the accompanying [technique](https://www.w3.org/WAI/WCAG21/Techniques/general/G183)

#### When not to use
WHen links are part of a recognizeable nteractive strcture, e.g. a menu, lists of links in footers that are clearly separated from non-clickable text, etc.

**Comment by anthonydillon** [comment]
It's clear, thank you @lyubomir-popov. Few questions?

Can we programmaticly provide the best varieant when we know the situation (via strip maybe)?

What do we do about situations like mega nav? There should be a version of this pattern that covers that situation.

Should be create a fresh modifier of the link pattern? Invert is clear of dark background but now we have more purpose which should be reflected in the class name.

**I don't want to end up with tailwind of link patterns, otherwise we will never have consistently**

**Comment by clagom** [comment]
That is also my concern, that's why we tried having a simpler solution that worked across all those situations, even if it _looked worse_ :)

**Comment by lyubomir-popov** [comment]
@anthonydillon @clagom 

> Can we programmaticly provide the best varieant when we know the situation (via strip maybe)?

Possibly - suru strips could force the colour to be white for example... But this feels like a precedent. Do we change button styling depending on the parent? It seems like an analogous situation, and the buttons have even more flavours.

> What do we do about situations like mega nav? There should be a version of this pattern that covers that situation.

We don't use it there (first bullet point in the "When not to use" section). We tried mixing links and text inside the navigation with @davegoddard42, and found out it quickly becomes confusing.

> Should be create a fresh modifier of the link pattern? Invert is clear of dark background but now we have more purpose 
which should be reflected in the class name.

Possibly... although this whole issue is about untangling the complicated issues around the accessibility of this specific component, I don't see why it needs a new name.

> _I don't want to end up with tailwind of link patterns, otherwise we will never have consistently_

I don't think this solution steers us towards the tailwind model. It is more like the buttons component - there's quite a few flavours there that fit specific use cases.

**Comment by lyubomir-popov** [comment]
Sorry after I wrote this I had an idea that could simplify things a bit - what if on backgrounds darker than #2d2d2d we simply use a link with a ".is-on-dark" class or something? This way the inverted link will only cover the scenario where the background is a gradient or image, and for the normal dark theme on solid background we just have a different blue that is accessible? The same way it is here on github:

![image](https://user-images.githubusercontent.com/2741678/159017163-db007257-7194-4f1f-bd5b-164e2f9b5535.png)


In this cas,e the updated rules would be:


#### When to use:
When links are surrounded by regular, non-clickable text on backgrounds lighter thant #2d2d2d, where it is impossible to obtain a blue colour that simultaneously has 3:1 contrast with surrounding white text, and at the same time, 4.5:1 contrast with the background as per WCAG rule [1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) and the accompanying [technique](https://www.w3.org/WAI/WCAG21/Techniques/general/G183)

Examples of this could be gradients, photographs, etc.:
![image](https://user-images.githubusercontent.com/2741678/159017639-ebdd3cfe-0f39-47ee-8b0d-5c566809a3ca.png)

#### When not to use
- In cases where the background can be set to a dark gray (`#2d2d2d` or darker). In this case, use the '.is-dark' modifier on a regulr link, e.g. `<a src='#' alt='' class='is-dark'>Link</a>`

- When the background does not allow 4.5:1 contrast with white text. In this case, consider changing the background or placing links on a surface that allows achieving the required contrast ratios.

### IsS-dark link
Then, for the `is-dark` link, there will be one exception:
- When links are part of a recognizeable interactive strcture, e.g. a menu, nav, a list of links in a site footer that are clearly separated from non-clickable text, etc.

**Comment by clagom** [comment]
+1 @lyubomir-popov I think this proposal simplifies the guidelines to a fewer use-cases, easier to apply.

**Comment by anthonydillon** [comment]
I am sorry @lyubomir-popov I am trying to wrap my head around the aspects of this proposal and failing. May you take me through it or provide pseudo CSS with comments for me?

**Comment by lyubomir-popov** [comment]
@anthonydillon something like this, maybe with a note in the docs that the two classes shouldn't be applied simultaneously:

```
    a.is-on-dark { color: $dark-theme__link-blue;}
    .p-inverted-link { text-decoration: underline; }
```

**Comment by anthonydillon** [comment]
After @lyubomir-popov kindly explained the proposal I am ok with the renaming of the classes but of course not set in stone and will need more thinking to ensure there are no issues introduced by the new class names.
